src/Tables.jl

```diff
@@ -8,98 +8,277 @@ if !hasmethod(getproperty, Tuple{Tuple, Int})
+| ---------------- | ------------------ | ----------------- |
+| `Tables.getcolumn(table, i::Int)` | getfield(table, i) | Retrieve a column by index |
+| `Tables.getcolumn(table, nm::Symbol)` | getproperty(table, nm) | Retrieve a column by name |
+| `Tables.columnnames(table)` | propertynames(table) | Return column names for a table as an indexable collection |
+| Optional methods | | |
+| `Tables.getcolumn(table, ::Type{T}, i::Int, nm::Symbol)` | Tables.getcolumn(table, nm) | Given a column eltype `T`, index `i`, and column name `nm`, retrieve the column. Provides a type-stable or even constant-prop-able mechanism for efficiency.
+
+Note that table sources shouldn't subtype `Columns`, as it is purely an interface type
+to help document the Tables.jl API. See the [`Tables.AbstractColumns`](@ref) type
+for a type to potentially subtype to gain useful default behaviors.
+"""
+abstract type Columns end
 
-So how does one go about satisfying the Tables.jl interface functions? It mainly depends on what you've already defined and the natural access patterns of your table:
+"""
+    Tables.AbstractColumns
+
+Abstract type provided to allow custom table types to inherit useful and required behavior. Note that this type
+is for convenience for table _source_ authors to provide useful default behavior to their `Columns` object,
+and not to be used or relied upon by sink authors to dispatch on; i.e. not all `Columns` objects will inherit
+from `Tables.AbstractColumns`.
+
+Interface definition:
+| Required Methods | Default Definition | Brief Description |
+| ---------------- | ------------------ | ----------------- |
+| `Tables.getcolumn(table, i::Int)` | getfield(table, i) | Retrieve a column by index |
+| `Tables.getcolumn(table, nm::Symbol)` | getproperty(table, nm) | Retrieve a column by name |
+| `Tables.columnnames(table)` | propertynames(table) | Return column names for a table as an indexable collection |
+| Optional methods | | |
+| `Tables.getcolumn(table, ::Type{T}, i::Int, nm::Symbol)` | Tables.getcolumn(table, nm) | Given a column eltype `T`, index `i`, and column name `nm`, retrieve the column. Provides a type-stable or even constant-prop-able mechanism for efficiency.
+
+While custom table types aren't required to subtype `Tables.AbstractColumns`, benefits of doing so include:
+  * Indexing interface defined (using `getcolumn`); i.e. `tbl[i]` will retrieve the column at index `i`
+  * Property access interface defined (using `columnnames` and `getcolumn`); i.e. `tbl.col1` will retrieve column named `col1`
+  * Iteration interface defined; i.e. `for col in table` will iterate each column in the table
+  * A default `show` method
+This allows a custom table type to behave as close as possible to a builtin `NamedTuple` of vectors object.
+"""
+abstract type AbstractColumns end
 
-First:
-* `Tables.istable(::Type{<:MyTable}) = true`: this provides an explicit affirmation that your type implements the Tables interface
+"""
+    Row
```

Comment:
I don't understand the reason for having this type. It's confusing to have both `Row` and `AbstractRow`, the two being... abstract. If the point is just to document the `Row` *trait*, then better just mention this in the manual (and in the `AbstractRow` docstring (which already says inheriting isn't needed, but documents the interface anyway).

Also:
```suggestion
    Tables.Row
```

Reply:
I figured the more docs the better, but yeah, maybe we only need to mention this in the docs.